File: Data Structures & Algorithms/max-water-container/submission-0.py
class Solution:
    def maxArea(self, heights: List[int]) -> int:
        left, right = 0, len(heights)-1
        best = 0
        while left < right:
            width = right - left
            height = min(heights[left], heights[right])
            best = max(best, width * height)

            if heights[left] < heights[right]:
                left = left + 1
            else:
                right = right - 1
            

        
        return best
        

File: Data Structures & Algorithms/binary-tree-diameter/submission-0.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {
    public int diameterOfBinaryTree(TreeNode root) {
        int[] res = new int[1];
        getDiameter(root, res);
        return res[0];
    }

    private int getDiameter(TreeNode root, int[] res) {
        if (root == null) {
            return 0;
        }
        int left = getDiameter(root.left, res);
        int right = getDiameter(root.right, res);
        res[0] = Math.max(res[0], left + right);
        return 1 + Math.max(left, right);
    }
}
